packages/core/src/browser/common-frontend-contribution.ts

```diff
@@ -354,16 +354,38 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
         this.updateThemeFromPreference('workbench.colorTheme');
         this.updateThemeFromPreference('workbench.iconTheme');
         this.preferences.onPreferenceChanged(e => {
-            if (e.preferenceName === 'workbench.editor.highlightModifiedTabs') {
-                this.updateStyles();
-            } else if (e.preferenceName === 'workbench.colorTheme' || e.preferenceName === 'workbench.iconTheme') {
-                this.updateThemeFromPreference(e.preferenceName);
+            switch (e.preferenceName) {
+                case 'workbench.editor.highlightModifiedTabs': {
+                    this.updateStyles();
+                    break;
+                }
+                case 'workbench.colorTheme':
+                case 'workbench.iconTheme': {
+                    this.updateThemeFromPreference(e.preferenceName);
+                    break;
+                }
+                case 'window.menuBarVisibility': {
+                    const { newValue } = e;
+                    const mainMenuId = 'main-menu';
+                    if (newValue === 'compact') {
+                        app.shell.leftPanelHandler.addTopMenu({
+                            id: mainMenuId,
+                            iconClass: 'codicon codicon-menu',
+                            title: 'Application Menu',
+                            menuPath: ['menubar'],
+                            order: 0,
+                        });
+                    } else {
+                        app.shell.leftPanelHandler.removeTopMenu(mainMenuId);
+                    }
+                    break;
+                }
```

Comment:
We should refactor and create a `onPreferenceChange` handler method so it will be easier for downstream applications to override if necessary (ex: if they want to handle the preference change event differently).

Reply:
I separated it to another protected function, is that what you meant?

Reply:
Correct, it makes it easier to extend if needed, and also cleans it up.